metabase/metabase
# Add custom warning message when leaving creating/editing a DB connection

Closes #33794

### Description

- includes [converting DatabaseEditApp.unit.spec.js to TypeScript](https://github.com/metabase/metabase/pull/34272/commits/f23d1565bb83fe46226370c13411c56d3f334fb9)
- similar to
  - #34084 
  - #34192
  - #33827


### How to verify

Prerequisites for all scenarios:

1. Log in as an admin
2. Open `/admin/databases`

**Custom warning modal is shown when leaving adding database with unsaved changes**

1. Click "Add database"
2. Change anything in the form
3. Click the back button in your browser or any link in page header

Expected result: custom warning modal **is shown**.

**Custom warning modal is not shown when leaving adding database with no changes**

1. Click "Add database"
2. (Optional) change something in the form and revert this change
3. Click the back button in your browser or any link in page header

Expected result: custom warning modal **is not shown**.

**Custom warning modal is shown when leaving editing database with unsaved changes**

1. Click "Sample database"
2. Change anything in the form
3. Click the back button in your browser or any link in page header

Expected result: custom warning modal **is shown**.

**Custom warning modal is not shown when leaving editing database with no changes**

1. Click "Sample database"
2. (Optional) change something in the form and revert this change
3. Click the back button in your browser or any link in page header

Expected result: custom warning modal **is not shown**.

**Custom warning modal is not shown when creating a database**

1. Click "Add database"
2. Connect a real db
3. Save
4. Wait for redirect to `/admin/databases`

Expected result: custom warning modal **is not shown**.

**Custom warning modal is not shown when updating a database**

1. Click "Sample database"
2. Change db name
3. Save
4. Wait for save to complete
5. Click the back button in your browser or any link in page header

Expected result: custom warning modal **is not shown**.

### Demo


https://github.com/metabase/metabase/assets/6830683/5bf49a8d-33c6-4ef8-beb4-7ac1cff17aac
